### PR TITLE
Feature: folder list and creation

### PR DESCRIPTION
**Pull request by nicorsm**
## What's new in this PR?

PoC for folder list (through `FolderPickerViewController`) and creation (through `FolderCreationController`).

### CODE REVIEW

**Comment by typfel** [comment]
The API for creating folders and retrieving folders is merged to develop so you can already integrate it.

File: Wire-iOS Tests/FolderCreationControllerSnapshotTests.swift [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    
+    func testForEditingTextField() {
+        
+        sut.loadViewIfNeeded()
+        sut.beginAppearanceTransition(false, animated: false)
+        sut.endAppearanceTransition()
+        
+        sut.viewDidAppear(false)
+        
+        verify(view: sut.view)
```
Comment (billypchan) [inline Wire-iOS Tests/FolderCreationControllerSnapshotTests.swift]: The header title is cropped, it needs more than one line.

File: Wire-iOS/Sources/UserInterface/Folders/FolderCreationController.swift [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    public init(conversation: ZMConversation) {
+        self.conversation = conversation
+        super.init(nibName: nil, bundle: nil)
+    }
+    
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    override public var prefersStatusBarHidden: Bool {
```
Comment (billypchan) [inline Wire-iOS/Sources/UserInterface/Folders/FolderCreationController.swift]: `public` is not needed

File: Wire-iOS Tests/FolderCreationControllerSnapshotTests.swift [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import XCTest
+@testable import Wire
+
+final class FolderCreationControllerSnapshotTests: CoreDataSnapshotTestCase {
```
Comment (billypchan) [inline Wire-iOS Tests/FolderCreationControllerSnapshotTests.swift]: You may try using new `SnapshotTesting` framework with the help of `CoreDataFixtureTestHelper` protocol. ref:`ConversationViewControllerSnapshotTests`

File: Wire-iOS.xcodeproj/xcshareddata/xcschemes/Wire-iOS.xcscheme [hunk 4]
```diff
@@ -120,7 +120,7 @@
          </CommandLineArgument>
          <CommandLineArgument
             argument = "-ZMBackendEnvironmentType staging"
-            isEnabled = "NO">
+            isEnabled = "YES">
```
Comment (billypchan) [inline Wire-iOS.xcodeproj/xcshareddata/xcschemes/Wire-iOS.xcscheme]: remember to undo this

File: Wire-iOS/Sources/UserInterface/ConversationList/Folders/CreateFolderViewController.swift [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import UIKit
+
+class CreateFolderViewController: UIViewController {
```
Comment (billypchan) [inline Wire-iOS/Sources/UserInterface/ConversationList/Folders/CreateFolderViewController.swift]: empty class?

File: Wire-iOS/Sources/UserInterface/Folders/FolderCreationController.swift [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+        case let .error(error):
+            print(error)
+        case let .valid(name):
+            let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
+            nameSection.resignFirstResponder()
+            folderName = trimmed
+            
+            if let folder = ZMUserSession.shared()?.conversationDirectory.createFolder(folderName) {
+                self.delegate?.folderController(self, didCreateFolder: folder)
+                self.navigationController?.popViewController(animated: true)
```
Comment (typfel) [inline Wire-iOS/Sources/UserInterface/Folders/FolderCreationController.swift]: We should not go back to list after creating a folder. We should dismiss the whole modal and move the conversation to the newly created folder.

File: Wire-iOS/Sources/UserInterface/Folders/FolderPickerViewController.swift [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    }
+    
+    private func pickFolder(_ folder: LabelType) {
+        
+        self.conversation.moveToFolder(folder)
+        self.delegate?.didPickFolder(folder, for: conversation)
+        self.dismissIfNeeded()
+    }
+    
+    private func handle(error: Error) {
```
Comment (typfel) [inline Wire-iOS/Sources/UserInterface/Folders/FolderPickerViewController.swift]: Can be deleted?

File: Wire-iOS/Sources/UserInterface/Folders/FolderPickerViewController.swift [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+        cell.title = item.name
+        cell.showCheckmark = false
+        cell.showSeparator = indexPath.row < (items.count - 1)
+        
+        return cell
+    }
+    
+    private func pickFolder(_ folder: LabelType) {
+        
+        self.conversation.moveToFolder(folder)
```
Comment (typfel) [inline Wire-iOS/Sources/UserInterface/Folders/FolderPickerViewController.swift]: This needs wrapped in `enqueueChanges`? But maybe we should perform the actual move in the delegate?

File: Wire-iOS/Sources/UserInterface/GroupDetails/ConversationActions/ConversationActionController.swift [hunk 9]
```diff
@@ -66,7 +66,7 @@ final class ConversationActionController {
         ZMUserSession.shared()?.enqueueChanges(block)
     }
     
-    func transitionToListAndEnqueue(_ block: @escaping () -> Void) {
+    internal func transitionToListAndEnqueue(_ block: @escaping () -> Void) {
```
Comment (typfel) [inline Wire-iOS/Sources/UserInterface/GroupDetails/ConversationActions/ConversationActionController.swift]: internal is the default.

File: Wire-iOS/Sources/UserInterface/Folders/FolderCreationController.swift [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+        UIApplication.shared.wr_updateStatusBarForCurrentControllerAnimated(animated)
+    }
+    
+    override public func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+        nameSection.becomeFirstResponder()
+    }
+    
+    private func setupViews() {
+        // TODO: if keyboard is open, it should scroll.
```
Comment (typfel) [inline Wire-iOS/Sources/UserInterface/Folders/FolderCreationController.swift]: @nicorsm what are you referring to here?

**Pull request merged**